Support configuring the bot from a Python module

RyanBalfanz opened the pull request:
This enables arbitrary Python code to configure 
the bot (e.g. using environment variables).

RyanBalfanz committed:
Support configuring the bot from a Python module

This enables arbitrary Python code to configure 
the bot (e.g. using environment variables).

tdeck reviewed (commented):
Thanks for bringing attention to this - the file config does seem to have its limitations and this is a nice clean PR. I really appreciate it!

Since we only have 3 variables to set though, I'm wondering if allowing the bot to be configured with either environment variables or explicit arguments would work for your use case. That seems like it would be easier for to deploy in a lot of environments.

If you need to integrate it into a larger Python application, I can make it a module that can be imported, and will run itself if started directly (the old `if __name__ == "__main__")`. It's sort of there already but doesn't look nice.

Sorry for bikeshedding on your PR. My hesitation stems from the fact that while rtmbot was written to dynamically pull in other modules, I never liked that design much and I've tried to chip away at it. If we're going to be invoking this from somewhere, it would be cleaner to just express it that way.

It occurs to me that this may be part of a project to get the slack token into Keywhiz, in which case it seems like the easiest thing would be to stick with the regular config file, but pop the whole thing into a keywhiz mount and set the path there.

RyanBalfanz commented:
I need this to be able to run on Heroku -- it has nothing to do with KeyWhiz. I haven't submitted a [PR for that](https://github.com/RyanBalfanz/teambot/tree/deploy-to-heroku-button) yet, because I'd like to get this merged in first.

What's missing from the Heroku PR is the environment variables in the `app.json`. I want to be able to deploy this quickly in a single click. The app owner will tell Heroku their `SLACK_TOKEN` at creation time, as well as set their preferred `TEAMBOT_SETTINGS_MODULE` (which does all the `os.getenv` stuff to get the token and other config.).

I really wanted to not muck with your existing stuff too much, so as to be backward compatible (esp. with regard to how the script is invoked). But, I hit a snag with that so I needed to introduce the magic `TEAMBOT_SETTINGS_MODULE` that cues the script how to configure itself in the absence of any arguments.

RyanBalfanz commented on rtmbot.py:
```diff
@@ -145,14 +147,42 @@ def main_loop():
+        metavar='module'
+    )
+
+    # Allows the bot to be run with zero arguments and use a Python module
```
@tdeck This is where the magic happens.

RyanBalfanz replied:
Since the group is mutually exclusive, we cannot set the defaults for both, or invoking the bot will be buggy. So, look for an environment variable named `TEAMBOT_SETTINGS_MODULE` to decide what to do.

RyanBalfanz commented on rtmbot.py:
```diff
@@ -145,14 +147,42 @@ def main_loop():
+    # Allows the bot to be run with zero arguments and use a Python module
+    # instead of a YAML file.
+    if os.getenv('TEAMBOT_SETTINGS_MODULE') is not None:
+        parser.set_defaults(settings=os.getenv('TEAMBOT_SETTINGS_MODULE'))
```
@tdeck Naming [borrowed from Django](https://docs.djangoproject.com/en/1.10/topics/settings/#envvar-DJANGO_SETTINGS_MODULE).

RyanBalfanz commented on rtmbot.py:
```diff
@@ -145,14 +147,42 @@ def main_loop():
+    elif args.settings:
+        s = importlib.import_module(args.settings)
+        get_s = functools.partial(getattr, s)
+        config = {
```
@tdeck I'm not super happy with this, i.e. explicitly listing the configs here.

Maybe something like `defaultSettings.update(userSettings)` would be nicer, but I wanted to keep it simple. But the defaults I've chosen are sane, and in line with your original design. Sadly, I didn't write tests to prove that. But I did test it well.

tdeck commented:
That Heroku work is really nice! I'd be happy to merge it and I appreciate it.

Keep in mind the original design of Teambot isn't too great, so it's all up for negotiation. I basically just pulled in rtmbot one 4th of July to make it work, and then progressively deleted things I didn't need to make it simpler over the last couple of years. I myself find the architecture of it a bit unintuitive. 

Your project seems perfectly reasonable to me, but if it meets your requirements I'd rather just have something that pulls the slack token and DEBUG straight from environment variables, and either dispenses with the config file entirely or leaves it in for backwards compatibility. That way you get a nice 12-factor app without having to load another module.  Not that your module loading code is hard to understand, but we're just pulling in constants from that module so it's just an indirection.

I should also apologize for taking so long to reply to these each time. I've had a lot going on at work lately, but my latency should be lower on the weekend.

RyanBalfanz commented:
This PR leaves it in for backwards compatibility. That was my original intent.

Usage of a Python module is triggered by one of:
 - `python rtmbot.py -s settings.module.path`
 - `python rtmbot.py --settings settings.module.path`
 - `TEAMBOT_SETTINGS_MODULE=settings.module.path python rtmbot.py`

Additionally, `TEAMBOT_SETTINGS_MODULE=settings.a python rtmbot.py --settings settings.b` will use `settings.b` (args trump env vars).

Running `rtmbot.py` with both `-s/--settings` with `-c/--config` is not allowed. Behavior when running `rtmbot.py` with no args is unchanged.

The original README is still technically correct, but I added here additional notes describing how to use a module path.

tdeck commented:
Yep I understand the intent, don't worry your code is pretty clear. I am
just wondering if there is any need for this indirection though a python
module, of if you'd be just as happy if I made a commitment where teambot
read the env vars directly. In other words, is there a use case I am
missing for importing settings from code?

On Mar 24, 2017 11:46 PM, "Ryan Balfanz" <notifications@github.com> wrote:

This PR leaves it in for backwards compatibility. That was my original
intent.

Usage of a Python module is triggered by one of:

   - python rtmbot.py -s settings.module.path
   - python rtmbot.py --settings settings.module.path
   - TEAMBOT_SETTINGS_MODULE=settings.module.path python rtmbot.py

Additionally, TEAMBOT_SETTINGS_MODULE=settings.a python rtmbot.py
--settings settings.b will use settings.b (args trump env vars).

The original README is still technically correct, but I added here
additional describing how to use a module path.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/tdeck/teambot/pull/6#issuecomment-289193261>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ACALLubbv_S09bPW4g_a4064JPEoWHh1ks5rpLhUgaJpZM4MnsOA>
.

RyanBalfanz commented:
Oh ok, I guess I probably don't understand what you're proposing. I guess if env vars were supported out of the box that would satisfy my Heroku requirements. But it sounds like you're suggesting removing the `rtmbot.conf` completely, which is of course not backward compatible. Some people might still like to have that. Am I misunderstanding?

RyanBalfanz commented:
Also, the settings from the module need not be constants. I think you may have suggested that before.

tdeck commented:
Oh I see. That was just an aside suggestion. I'm not really planning to do
that at first, although the migration path isn't unduly onerous I would
say. I will revisit tomorrow.

On Mar 24, 2017 11:58 PM, "Ryan Balfanz" <notifications@github.com> wrote:

> Oh ok, I guess I probably don't understand what you're proposing. I guess
> if env vars were supported out of the box that would satisfy my Heroku
> requirements. But it sounds like you're suggesting removing the
> rtmbot.conf completely, which is of course not backward compatible. Some
> people might still like to have that. Am I misunderstanding?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tdeck/teambot/pull/6#issuecomment-289193777>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACALLhTX3lsJ340iVKruFxiln5L1y2-Gks5rpLsdgaJpZM4MnsOA>
> .
>

RyanBalfanz commented:
> Create a settings module and set `TEAMBOT_SETTINGS_MODULE` in your environment to the Python path of the module. Once simple way is to `cp settings.py.example settings.py`, then edit settings.py as needed.

Maybe this would be more clear if I also mentioned that instead of (or in addition to) an env var, you can swap the existing `-c` arg with a `-s` and give it a Python path to a module.

Does that help any?

RyanBalfanz committed:
Add command argument to display settings config and exit

RyanBalfanz committed:
Test parse_args using doctest


```
(env) ➜  teambot git:(settings-module-config) ✗ make test
python -m doctest -v rtmbot.py
Trying:
    from test.test_support import EnvironmentVarGuard
Expecting nothing
ok
Trying:
    env = EnvironmentVarGuard()
Expecting nothing
ok
Trying:
    env.unset('TEAMBOT_SETTINGS_MODULE')
Expecting nothing
ok
Trying:
    parse_args([])
Expecting:
    Namespace(config='rtmbot.conf', diffsettings=False, settings=None)
ok
Trying:
    parse_args(["--diffsettings"])
Expecting:
    Namespace(config='rtmbot.conf', diffsettings=True, settings=None)
ok
Trying:
    parse_args(["--config", 'teambotconfig.conf'])
Expecting:
    Namespace(config='teambotconfig.conf', diffsettings=False, settings=None)
ok
Trying:
    env.set('TEAMBOT_SETTINGS_MODULE', 'settings')
Expecting nothing
ok
Trying:
    parse_args([])
Expecting:
    Namespace(config=None, diffsettings=False, settings='settings')
ok
Trying:
    parse_args(["--diffsettings"])
Expecting:
    Namespace(config=None, diffsettings=True, settings='settings')
ok
Trying:
    parse_args(["--settings", 'teambotconfig'])
Expecting:
    Namespace(config=None, diffsettings=False, settings='teambotconfig')
ok
17 items had no tests:
    rtmbot
    rtmbot.Plugin
    rtmbot.Plugin.__init__
    rtmbot.Plugin.do
    rtmbot.Plugin.do_output
    rtmbot.RtmBot
    rtmbot.RtmBot.__init__
    rtmbot.RtmBot.autoping
    rtmbot.RtmBot.connect
    rtmbot.RtmBot.input
    rtmbot.RtmBot.load_plugins
    rtmbot.RtmBot.output
    rtmbot.RtmBot.start
    rtmbot.UnknownChannel
    rtmbot.dbg
    rtmbot.get_config
    rtmbot.main_loop
1 items passed all tests:
  10 tests in rtmbot.parse_args
10 tests in 18 items.
10 passed and 0 failed.
Test passed.
```

RyanBalfanz committed:
Add Travis-CI config

RyanBalfanz committed:
Add Travis-CI build status image

TODO: Change this back to tdeck/teambot@master once merged

tdeck commented:
Sorry I guess I must not be expressing myself very clearly. To put it bluntly, I don't like this design. The goal is to be able to read the config values from env vars - that makes sense and I think we should just do that. This thing of specifying a Python module path on the command line, dynamically loading that module, and pulling constants out of it, then having a separate module that sets the constants based on environment variables, feels overly convoluted and I don't see the benefit of doing it this way.

I think instead we should just do something like I have in this working branch:
https://github.com/tdeck/teambot/commit/58f242733787f97a46cffe20930c3811d8260cbe

However, I don't mean any disrespect to your work. I just believe that if we don't need t he flexibility of being able to load an arbitrary module, we shouldn't add it.

RyanBalfanz commented:
That's fine. Though I'm not sure what you mean by "then have a separate
module". There's only one module.

Yes, my only real ask if allowing it to be configured 100% from the
environment. I designed it this way to try and match the original design
(specifically, the args usage). But to enable the command to be run without
any modification at all (my intent was to not break existing installs or
completely rewrite your project) forced needing a way to cue the code to
load a module and not a .yml file. Actually, I hoped that's by keeping the
original usage intact you'd be _more_ likely to approve my changes and
incorporate them. :-/

I'm on mobile now and can't take a good look at your branch now but I will.
If that work does what I need that's great. What will it take to finish
that work and get it merged? And when can I roughly expect it might be
completed?

I could absolutely use this tomorrow. But I can wait too. What can I do, if
anything, to help get this over the finish line in the next few days?
On Sat, Mar 25, 2017 at 11:02 PM Troy Deck <notifications@github.com> wrote:

> Sorry I guess I must not be expressing myself very clearly. To put it
> bluntly, I don't like this design. The goal is to be able to read the
> config values from env vars - that makes sense and I think we should just
> do that. This thing of specifying a Python module path on the command line,
> dynamically loading that module, and pulling constants out of it, then
> having a separate module that sets the constants based on environment
> variables, feels overly convoluted and I don't see the benefit of doing it
> this way.
>
> I think instead we should just do something like I have in this working
> branch:
> 58f2427
> <https://github.com/tdeck/teambot/commit/58f242733787f97a46cffe20930c3811d8260cbe>
>
> However, I don't mean any disrespect to your work. I just believe that if
> we don't need t he flexibility of being able to load an arbitrary module,
> we shouldn't add it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tdeck/teambot/pull/6#issuecomment-289260721>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIInrFgBqPantl0r7uC6gkGFsOaESTUks5rpf93gaJpZM4MnsOA>
> .
>

tdeck commented:
I'm working on it in my branch. Also going to add the ability to specify the path of the DB file, as I think that might be convenient. Need to test my changes to make sure they're sane.

tdeck commented:
Closing this since we ended up going a different route to support env vars. Thanks for writing it

RyanBalfanz commented:
@tdeck If I pulled out the Travis stuff from this branch into a new one, do you think you'd be open to merging it? That, plus the docker stuff, will help me make additional contributions in the future.

tdeck commented:
It would be good to have CI tests; I had thought about it but hadn't done it because it's going to be a real pain to mock out the API surface, but if you're willing to write some that would be awesome. I can help if I have time, in the near future I may not but I'll review your PR.
